aggregator/src/core.rs

```diff
@@ -62,3 +82,674 @@ pub(crate) fn extract_accumulators_and_proof(
+                    &mut region,
+                    &num_of_valid_snarks_cell[0],
+                    &const32,
+                    &mut offset,
+                )?;
+
+                // sanity check
+                assert_exist(
+                    &data_hash_inputs,
+                    &hash_input_len_cells[MAX_AGG_SNARKS * 2 + 3],
+                    &hash_input_len_cells[MAX_AGG_SNARKS * 2 + 4],
+                    &hash_input_len_cells[MAX_AGG_SNARKS * 2 + 5],
+                );
+
+                log::trace!("data_hash_inputs: {:?}", data_hash_inputs.value());
+                log::trace!(
+                    "candidate 1: {:?}",
+                    hash_input_len_cells[MAX_AGG_SNARKS * 2 + 3].value()
+                );
+                log::trace!(
+                    "candidate 2: {:?}",
+                    hash_input_len_cells[MAX_AGG_SNARKS * 2 + 4].value()
+                );
+                log::trace!(
+                    "candidate 3: {:?}",
+                    hash_input_len_cells[MAX_AGG_SNARKS * 2 + 5].value()
+                );
+
+                let mut data_hash_inputs_rec = rlc_config.mul(
+                    &mut region,
+                    &hash_input_len_cells[MAX_AGG_SNARKS * 2 + 3],
+                    &flag1,
+                    &mut offset,
+                )?;
+                data_hash_inputs_rec = rlc_config.mul_add(
+                    &mut region,
+                    &hash_input_len_cells[MAX_AGG_SNARKS * 2 + 4],
+                    &flag2,
+                    &data_hash_inputs_rec,
+                    &mut offset,
+                )?;
+                data_hash_inputs_rec = rlc_config.mul_add(
+                    &mut region,
+                    &hash_input_len_cells[MAX_AGG_SNARKS * 2 + 5],
+                    &flag3,
+                    &data_hash_inputs_rec,
+                    &mut offset,
+                )?;
+
+                // sanity check
+                assert_equal(&data_hash_inputs, &data_hash_inputs_rec);
+                region.constrain_equal(data_hash_inputs.cell(), data_hash_inputs_rec.cell())?;
+                log::trace!("rlc chip uses {} rows", offset);
+                Ok(num_of_valid_snarks_cell)
+            },
+        )
+        .map_err(|e| Error::AssertionFailure(format!("aggregation: {e}")))?;
+    Ok(num_of_valid_snarks_cell[0].clone())
+}
+
+/// generate a string of binary cells indicating
+/// if the i-th chunk is a valid chunk
+pub(crate) fn chunk_is_valid(
+    rlc_config: &RlcConfig,
+    region: &mut Region<Fr>,
+    num_of_valid_chunks: &AssignedCell<Fr, Fr>,
+    offset: &mut usize,
+) -> Result<[AssignedCell<Fr, Fr>; MAX_AGG_SNARKS], halo2_proofs::plonk::Error> {
+    let mut res = vec![];
+
+    for i in 0..MAX_AGG_SNARKS {
+        let value = rlc_config.load_private(region, &Fr::from(i as u64), offset)?;
+        let is_valid = rlc_config.is_smaller_than(region, &value, num_of_valid_chunks, offset)?;
+        res.push(is_valid);
```

Comment:
First of all, the `is_smaller_than` rely on some assumption that both operands are small, but both the `num_of_valid_chunks` and `value` are private cells.
The circuit itself does not enforce any range constraints on them. 
Secondly, I think we should test if a chunk is padding by checking if the chunk data hash equals to `keccak("")`

Reply:
> First of all, the is_smaller_than rely on some assumption that both operands are small, but both the num_of_valid_chunks and value are private cells.

- num_of_valid_chunks is a public input https://github.com/scroll-tech/zkevm-circuits/blob/static-proof-aggregation/aggregator/src/aggregation/circuit.rs#L380-L386
- "rlc_config.load_private(region, &Fr::from(i as u64), offset)?;" missed a copy constraint from fixed column. should have loaded a constant 1 and accumulate from there. will fix

Reply:
also a random thought: would a real trace have an empty data field?

Reply:
A real trace will not have `data_hash = keccak([])` as it at least has one block context.